presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java

```diff
@@ -201,13 +203,15 @@ public List<Supplier<PageProjectionWithOutputs>> compileProjections(
         if (isOptimizeCommonSubExpression) {
             ImmutableList.Builder<Supplier<PageProjectionWithOutputs>> pageProjections = ImmutableList.builder();
             ImmutableMap.Builder<RowExpression, Integer> expressionsWithPositionBuilder = ImmutableMap.builder();
+            Set<RowExpression> expressionCandidates = new HashSet<>();
             for (int i = 0; i < projections.size(); i++) {
                 RowExpression projection = projections.get(i);
-                if (projection instanceof ConstantExpression || projection instanceof InputReferenceExpression) {
+                if (projection instanceof ConstantExpression || projection instanceof InputReferenceExpression || expressionCandidates.contains(projection)) {
```

Comment:
Can you add a comment about why we exclude duplicate projections?

Reply:
Done.

Reply:
Instead of just commenting what it will do, can you also comment the intent that we don't do CSE for duplicate expressions? How hard it is to do CSE in case of duplicate expressions as well?

Reply:
> Instead of just commenting what it will do, can you also comment the intent that we don't do CSE for duplicate expressions? 

I'm not sure but my guess is that it's just a rare corner case which is not considered during implementation. There is no particular reason why it's intrinsically hard to implement.

> How hard it is to do CSE in case of duplicate expressions as well?

I think the biggest challenge is that we need to refactor the CSE code a bit. Consider that this case is rare (we hit this single issue until now), refactor the code looks like not a good tradeoff.

Reply:
@feilong-liu can you update the comment to: 

`Duplicate expressions are not expected here in general due to <...>, hence we skip CSE for them to allow for a simpler implementation`

Reply:
Updated the comments.